components/server/src/auth/host-context-provider-impl.ts

```diff
@@ -32,33 +34,41 @@ export class HostContextProviderImpl implements HostContextProvider {
 
     async init() {
         await this.ensureInitialized();
+
+        // schedule periodic update of dynamic hosts
+        const updateDynamicHostsJob = repeat(async () => {
+            const span = TraceContext.startSpan("updateDynamicHosts");
+            try {
+                await this.updateDynamicHosts({span});
+            } catch (error) {
+                log.error(`Failed to update dynamic hosts.`, error);
+            } finally {
+                span.finish();
+            }
+        }, 10000);
+        this.disposables.push(updateDynamicHostsJob);
     }
 
+    /** ensures we're not running this multiple times in parallel */
+    protected initializationRunning = false;
     protected initialized = false;
     protected async ensureInitialized() {
-        if (this.initialized) {
+        if (this.initialized || this.initializationRunning) {
             return;
         }
-        this.createFixedHosts();
-
         try {
-            await this.updateDynamicHosts({ });
-        } catch (error) {
-            log.error(`Failed to update dynamic hosts.`, error);
-        }
+            this.initializationRunning = true;
+            this.createFixedHosts();
```

Comment:
@AlexTugarev Note that we do not handle the error case here, but leave it in `this.initialized = false` state, keeping the current logic. Does that really make sense? :thinking: